File: android-app/app/src/main/java/com/phonecam/streamer/rewards/RewardManager.kt
package com.phonecam.streamer.rewards

import org.json.JSONObject

/**
 * Rewarded-ads credit engine.
 *
 * This is a 1:1 port of the reference implementation and its test suite at
 * reward_engine/reward_manager.py (see that file's docstring for the design
 * rationale: watching a full batch of [RewardConfig.adsPerReward] ads adds
 * one reward's worth of real time on top of whatever premium expiry is
 * already in effect — not a balance that only drains while streaming, and
 * not one hour per single ad). Any behavioral change here should be made in
 * the Python reference first, where it can be tested without a device/emulator,
 * then mirrored here.
 */
data class RewardConfig(
    // Alpha/Beta rate: 3 ads = 12h of Pro (was 1h) — see AppPhase's doc for
    // why there's no paid tier at all right now, making the ad reward the
    // only way in and worth making generous.
    val secondsPerReward: Double = 12 * 3600.0,
    val adsPerReward: Int = 3,
    // Kept at 8x a single reward, same ratio as before the 1h->12h bump
    // (was 8x 1h = 8h), so "how many batches can stack" is unchanged even
    // though each batch is now worth much more.
    val maxBalanceSeconds: Double = 8 * (12 * 3600.0),
    val freeQuality: String = "1080p60",
    val premiumQuality: String = "4k60",
) {
    init {
        require(secondsPerReward > 0) { "secondsPerReward must be positive" }
        require(adsPerReward > 0) { "adsPerReward must be positive" }
        require(maxBalanceSeconds >= secondsPerReward) { "maxBalanceSeconds must be >= secondsPerReward" }
    }
}

data class StreamProfile(
    val quality: String,
    val watermark: Boolean,
    val adsEnabled: Boolean,
    val premiumActive: Boolean,
    val balanceSeconds: Double,
)

class RewardManager(
    val config: RewardConfig = RewardConfig(),
    premiumExpiresAt: Double = 0.0,
    adsWatchedInBatch: Int = 0,
) {
    init {
        require(premiumExpiresAt >= 0) { "premiumExpiresAt cannot be negative" }
        require(adsWatchedInBatch in 0 until config.adsPerReward) { "adsWatchedInBatch must be in [0, adsPerReward)" }
    }

    /** Epoch millis (as a Double, matching the Python reference's epoch seconds*1000) when premium runs out. 0 = never earned. */
    var premiumExpiresAt: Double = premiumExpiresAt
        private set

    var adsWatchedInBatch: Int = adsWatchedInBatch
        private set

    fun balanceSeconds(nowMs: Double = System.currentTimeMillis().toDouble()): Double =
        maxOf(0.0, (premiumExpiresAt - nowMs) / 1000.0)

    /**
     * Record one completed rewarded ad. Returns true iff this ad completed a
     * batch of [RewardConfig.adsPerReward] and extended the premium expiry.
     *
     * The extension stacks on whichever is later — "now" or the current
     * expiry — so watching another batch while premium is still active adds
     * a full extra hour on top instead of overwriting the remaining time.
     */
    fun creditAdWatch(nowMs: Double = System.currentTimeMillis().toDouble()): Boolean {
        adsWatchedInBatch += 1
        if (adsWatchedInBatch < config.adsPerReward) return false

        adsWatchedInBatch = 0
        val base = maxOf(nowMs, premiumExpiresAt)
        val extended = base + config.secondsPerReward * 1000.0
        val cap = nowMs + config.maxBalanceSeconds * 1000.0
        premiumExpiresAt = minOf(extended, cap)
        return true
    }

    fun isPremiumActive(nowMs: Double = System.currentTimeMillis().toDouble()): Boolean =
        premiumExpiresAt > nowMs

    fun currentProfile(nowMs: Double = System.currentTimeMillis().toDouble()): StreamProfile {
        val premium = isPremiumActive(nowMs)
        return StreamProfile(
            quality = if (premium) config.premiumQuality else config.freeQuality,
            watermark = !premium,
            adsEnabled = !premium,
            premiumActive = premium,
            balanceSeconds = balanceSeconds(nowMs),
        )
    }

    /** Serialize for persistence (SharedPreferences). */
    fun toJson(): JSONObject = JSONObject().apply {
        put("premium_expires_at", premiumExpiresAt)
        put("ads_watched_in_batch", adsWatchedInBatch)
        put(
            "config",
            JSONObject().apply {
                put("seconds_per_reward", config.secondsPerReward)
                put("ads_per_reward", config.adsPerReward)
                put("max_balance_seconds", config.maxBalanceSeconds)
                put("free_quality", config.freeQuality)
                put("premium_quality", config.premiumQuality)
            },
        )
    }

    companion object {
        fun fromJson(json: JSONObject): RewardManager {
            val cfgJson = json.optJSONObject("config")
            val defaults = RewardConfig()
            val config = RewardConfig(
                secondsPerReward = cfgJson?.optDouble("seconds_per_reward", defaults.secondsPerReward) ?: defaults.secondsPerReward,
                adsPerReward = cfgJson?.optInt("ads_per_reward", defaults.adsPerReward) ?: defaults.adsPerReward,
                maxBalanceSeconds = cfgJson?.optDouble("max_balance_seconds", defaults.maxBalanceSeconds) ?: defaults.maxBalanceSeconds,
                freeQuality = cfgJson?.optString("free_quality", defaults.freeQuality) ?: defaults.freeQuality,
                premiumQuality = cfgJson?.optString("premium_quality", defaults.premiumQuality) ?: defaults.premiumQuality,
            )
            // A corrupted/edited save file shouldn't crash the app on load.
            val adsWatched = json.optInt("ads_watched_in_batch", 0).let {
                if (it in 0 until config.adsPerReward) it else 0
            }
            return RewardManager(config, json.optDouble("premium_expires_at", 0.0), adsWatched)
        }
    }
}

File: reward_engine/reward_manager.py
"""Rewarded-ads credit engine.

Core economy for the app's free tier: users unlock a temporary "premium"
window (no watermark, no ads, max quality) by watching rewarded ads.

Premium time is a wall-clock expiry, not a balance that drains while
streaming: watching a full batch of `ads_per_reward` ads (default 3) adds
one reward's worth of real time on top of whatever expiry is already in
effect (or on top of "now" if premium isn't currently active), so back-to-back
batches stack instead of resetting. This means premium keeps counting down
in the background even when the user isn't actively streaming — the whole
point of "3 ads bought you an hour" is that it's an hour, not "an hour of
streaming spread across however many days you get around to it."

This module has no Android/UI dependencies on purpose: it is the reference
implementation. The Kotlin port in
android-app/app/src/main/java/com/phonecam/streamer/rewards/RewardManager.kt
mirrors this logic 1:1 so the economy can be tuned and tested here first.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class RewardConfig:
    """Tunable knobs for the ad-credit economy.

    Alpha/Beta defaults reflect "3 ads = 12 hours" (was 1 hour): there is no
    paid Pro tier at all while the app is in Alpha/Beta (see the Android
    app's AppPhase.kt), so the ad reward is the only way in and is worth
    making generous. max_balance_seconds is kept at 8x a single reward —
    the same "how many batches can stack" ratio as before the 1h->12h bump
    (was 8x 1h = 8h) — even though each batch is now worth much more.
    """

    seconds_per_reward: float = 12 * 3600.0
    ads_per_reward: int = 3
    max_balance_seconds: float = 8 * (12 * 3600.0)
    free_quality: str = "1080p60"
    premium_quality: str = "4k60"

    def __post_init__(self) -> None:
        if self.seconds_per_reward <= 0:
            raise ValueError("seconds_per_reward must be positive")
        if self.ads_per_reward <= 0:
            raise ValueError("ads_per_reward must be positive")
        if self.max_balance_seconds < self.seconds_per_reward:
            raise ValueError("max_balance_seconds must be >= seconds_per_reward")


@dataclass(frozen=True)
class StreamProfile:
    """The effective capabilities applied to the current/next stream frame."""

    quality: str
    watermark: bool
    ads_enabled: bool
    premium_active: bool
    balance_seconds: float


class RewardManager:
    """Tracks the user's ad-watch batch progress and premium expiry.

    Usage pattern:
        rm = RewardManager(config)
        completed = rm.credit_ad_watch()   # user finished ONE rewarded ad
        if completed:
            ...                            # that ad completed a batch — premium extended
        profile = rm.current_profile()
    """

    def __init__(
        self,
        config: Optional[RewardConfig] = None,
        premium_expires_at: float = 0.0,
        ads_watched_in_batch: int = 0,
    ) -> None:
        self.config = config or RewardConfig()
        if premium_expires_at < 0:
            raise ValueError("premium_expires_at cannot be negative")
        if not (0 <= ads_watched_in_batch < self.config.ads_per_reward):
            raise ValueError("ads_watched_in_batch must be in [0, ads_per_reward)")
        self._premium_expires_at = premium_expires_at
        self._ads_watched_in_batch = ads_watched_in_batch

    @property
    def ads_watched_in_batch(self) -> int:
        return self._ads_watched_in_batch

    @property
    def premium_expires_at(self) -> float:
        return self._premium_expires_at

    def balance_seconds(self, now: Optional[float] = None) -> float:
        now = time.time() if now is None else now
        return max(0.0, self._premium_expires_at - now)

    def credit_ad_watch(self, now: Optional[float] = None) -> bool:
        """Record one completed rewarded ad. Returns True iff this ad completed
        a batch of `ads_per_reward` and extended the premium expiry.

        The extension stacks on whichever is later — "now" or the current
        expiry — so watching another batch while premium is still active adds
        a full extra hour on top instead of overwriting the remaining time.
        Capped so a user can't binge-watch their way to weeks of premium.
        """
        now = time.time() if now is None else now
        self._ads_watched_in_batch += 1
        if self._ads_watched_in_batch < self.config.ads_per_reward:
            return False

        self._ads_watched_in_batch = 0
        base = max(now, self._premium_expires_at)
        extended = base + self.config.seconds_per_reward
        cap = now + self.config.max_balance_seconds
        self._premium_expires_at = min(extended, cap)
        return True

    def is_premium_active(self, now: Optional[float] = None) -> bool:
        now = time.time() if now is None else now
        return self._premium_expires_at > now

    def current_profile(self, now: Optional[float] = None) -> StreamProfile:
        now = time.time() if now is None else now
        premium = self.is_premium_active(now)
        return StreamProfile(
            quality=self.config.premium_quality if premium else self.config.free_quality,
            watermark=not premium,
            ads_enabled=not premium,
            premium_active=premium,
            balance_seconds=self.balance_seconds(now),
        )

    def to_dict(self) -> dict:
        """Serialize for persistence (e.g. SharedPreferences / a JSON file)."""
        return {
            "premium_expires_at": self._premium_expires_at,
            "ads_watched_in_batch": self._ads_watched_in_batch,
            "config": {
                "seconds_per_reward": self.config.seconds_per_reward,
                "ads_per_reward": self.config.ads_per_reward,
                "max_balance_seconds": self.config.max_balance_seconds,
                "free_quality": self.config.free_quality,
                "premium_quality": self.config.premium_quality,
            },
        }

    @classmethod
    def from_dict(cls, data: dict) -> "RewardManager":
        cfg_data = data.get("config", {})
        config = RewardConfig(
            seconds_per_reward=cfg_data.get(
                "seconds_per_reward", RewardConfig.seconds_per_reward
            ),
            ads_per_reward=cfg_data.get("ads_per_reward", RewardConfig.ads_per_reward),
            max_balance_seconds=cfg_data.get(
                "max_balance_seconds", RewardConfig.max_balance_seconds
            ),
            free_quality=cfg_data.get("free_quality", RewardConfig.free_quality),
            premium_quality=cfg_data.get("premium_quality", RewardConfig.premium_quality),
        )
        ads_watched = data.get("ads_watched_in_batch", 0)
        if not (0 <= ads_watched < config.ads_per_reward):
            ads_watched = 0
        return cls(
            config=config,
            premium_expires_at=data.get("premium_expires_at", 0.0),
            ads_watched_in_batch=ads_watched,
        )
